### DISCUSSION TITLE
Change Friendly Name from function?

### DISCUSSION THREAD

**Discussion by woggy81** [zachowj/hass-node-red#80] [Q&A] [answered]
Hi,

I want to know if it's possible to change the "Home Assistant Config (optional) Name" from a function. I know that it will create a entity ID the first time and that's good. But I have all my alarm zones mapped (zone 1-40) and when I change the name in my alarm panel i would be great if they change in HA to.

**Comment by zachowj** [comment, marked as answer]
You could format a `msg` to pass to an API node to update it in HA. It won't change the value in the entity node but if you trigger API node every time the entity node changes then it should stay synced.

```js
msg.payload = { 
  data: {
      type: "nodered/discovery",
      server_id: "aaa.123", // the node id from the server config
      node_id: "xyz.983", // the node id of the entity node
      component: "sensor", //  or binary_sensor, switch
      config: {
          name: "new friendly name"
      }
  }
};
```

**Reply by woggy81** [reply]
got it to work with sensor but not switch, I have no idea why =)

**Reply by bertbaron** [reply]
Could someone please elaborate on this example? I'm trying to change the icon of an entity but I can't get anything to work. As I can't find much documentation about using the API node nor using it with 'nodered/discovery' I'm just guessing how to use it with no luck so far.
I have at least the following concrete questions:
- How to find the 'server_id'?
- How to find the node id of the entity node? (is 'entity node' referring to the 'entity node' node type here?)
- How to configure the 'API node'? Should I just use the above (JSONata: payload) as data? Or payload.data? Or something else?

**Reply by zachowj** [reply]
> * How to find the 'server_id'?

Open a server config either via the configuration nodes tab or through an HA node editor. Then in the information tab look for the node info.

![image](https://user-images.githubusercontent.com/37859597/166510141-09683dd8-0e84-45e6-a32d-ecd8cc221936.png)

> * How to find the node id of the entity node? (is 'entity node' referring to the 'entity node' node type here?)

Same as above after selecting the entity node look in the information tab.

> * How to configure the 'API node'? Should I just use the above (JSONata: payload) as data? Or payload.data? Or something else?

just the `payload.data` part

**Reply by bertbaron** [reply]
Thanks. I think I tried it that way but I'll try again.

**Reply by bertbaron** [reply]
Unfortunately I keep getting the error `[error] [ha-api:MyNode] path.charAt is not a function` from the API node. I can't see where this error is coming from. I do see a 'path' config variable in [the source](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/blob/main/src/nodes/api/editor.ts) which may be the variable referred to by the error. I'm not sure how to debug or config its value though.

**Reply by zachowj** [reply]
If I had to guess without seeing what you had in the data field is that your JSONata is invalid.

**Reply by bertbaron** [reply]
Sorry, I should have provided more information. I've got it working with JSON. JSONata doesn't seem to work though, no matter what I enter as JSONata expression, it always results in the error `path.charAt is not a function`. So I think something is broken there. I can work with json however by simply creating multiple API Nodes, each with the desired values.

In case someone wants to investigate the issue further hereby some more information:

HA Core version: 2022.5.1 (but issue observed already with 2022.4.x)
Node Red add-on version: 11.1.2

JSON data that works: 
 ```json
 {
   "data": {
     "type": "nodered/discovery",
     "server_id": "cd12362b.2d1aa8",
     "node_id": "518a70e95cf800d6",
     "component": "sensor",
     "config": {
       "name": "Test123",
       "icon": "mdi:account"
     }
   }
 }
 ```
 Using the above as literal JSONata gives the mentioned error. But also literal JSONata `{}` or any other value I tried gives the error.

**Reply by bertbaron** [reply]
Sorry, the above is not entirely correct.  It seems like the input message resulted in the error even though I didn't use it in the JSONata expression. What works is the message:
```json
{
  "payload": {
    "type": "nodered/discovery",
    "server_id": "cd12362b.2d1aa8",
    "node_id": "518a70e95cf800d6",
    "component": "sensor",
    "config": {
      "name": "Testjes",
      "icon": "mdi:account"
    }
  }
}
```
with JSONata: `payload` (and other expressions give reasonable errors)

The problem was that I provided the message:
```json
{
  "payload": {
    "data": {
      "type": "nodered/discovery",
      "server_id": "cd12362b.2d1aa8",
      "node_id": "518a70e95cf800d6",
      "component": "sensor",
      "config": {
        "name": "Testjes",
        "icon": "mdi:account"
      }
    }
  }
}
```
To my surprise this results in the `path.charAt is not a function` regardless of the JSONata expression I use, including `payload.data`.
